File: zad_5.go
package main

import (
 "fmt"
 "math"
)

// Вычисляет температуру в момент времени t по закону Ньютона
func modelTemperature(T0, Ts, r, t float64) float64 {
 return Ts + (T0-Ts)*math.Exp(-r*t)
}

// Метод наименьших квадратов для линейной аппроксимации: T(t) = a*t + b
func linearFit(x, y []float64) (a, b float64) {
 n := float64(len(x))
 var sumX, sumY, sumXY, sumX2 float64

 for i := range x {
  sumX += x[i]
  sumY += y[i]
  sumXY += x[i] * y[i]
  sumX2 += x[i] * x[i]
 }

 denominator := n*sumX2 - sumX*sumX
 if denominator == 0 {
  return 0, 0
 }

 a = (n*sumXY - sumX*sumY) / denominator
 b = (sumY - a*sumX) / n
 return
}

// Вычисляет коэффициент корреляции между x и y
func correlation(x, y []float64) float64 {
 n := len(x)
 if n == 0 {
  return 0
 }

 var sumX, sumY float64
 for i := 0; i < n; i++ {
  sumX += x[i]
  sumY += y[i]
 }
 meanX := sumX / float64(n)
 meanY := sumY / float64(n)

 var num, dx2, dy2 float64
 for i := 0; i < n; i++ {
  dx := x[i] - meanX
  dy := y[i] - meanY
  num += dx * dy
  dx2 += dx * dx
  dy2 += dy * dy
 }

 if dx2 == 0 || dy2 == 0 {
  return 0
 }
 return num / math.Sqrt(dx2*dy2)
}

// Безопасный ввод положительного числа
func readPositive(prompt string) float64 {
 var val float64
 for {
  fmt.Print(prompt)
  _, err := fmt.Scan(&val)
  if err == nil && val > 0 {
   return val
  }
  fmt.Println("Ошибка: введите положительное число.")
 }
}

func main() {
 var T0, Ts, r, totalTime, dt float64
 var timePoints, tempPoints []float64

 // Ввод начальных данных
 fmt.Print("Начальная температура кофе (°C): ")
 fmt.Scan(&T0)

 fmt.Print("Температура окружающей среды (°C): ")
 fmt.Scan(&Ts)

 r = readPositive("Коэффициент охлаждения (1/мин): ")
 totalTime = readPositive("Общее время моделирования (мин): ")
 dt = readPositive("Шаг по времени (мин): ")

 // Вывод шапки таблицы
 fmt.Println("\nМоделирование остывания кофе (метод Ньютона)")
 fmt.Println("----------------------------------------------")
 fmt.Printf("%10s | %15s\n", "Время", "Температура")
 fmt.Println("----------------------------------------------")

 // Расчёт температуры на каждом шаге
 for t := 0.0; t <= totalTime+1e-9; t += dt {
  T := modelTemperature(T0, Ts, r, t)
  timePoints = append(timePoints, t)
  tempPoints = append(tempPoints, T)
  fmt.Printf("%10.2f | %15.2f\n", t, T)
 }
 fmt.Println("----------------------------------------------")

 // Аппроксимация и корреляция
 a, b := linearFit(timePoints, tempPoints)
 rVal := correlation(timePoints, tempPoints)

 // Вывод уравнения и корреляции
 fmt.Printf("\nЛинейная аппроксимация: T(t) = %.4f * t + %.4f\n", a, b)
 fmt.Printf("Коэффициент корреляции: r = %.4f\n", rVal)
}

File: zad_5.cpp
#include <iostream>     // Стандартный ввод/вывод
#include <iomanip>      // Для управления форматированием чисел
#include <cmath>        // Для математических функций (exp, sqrt)
#include <vector>       // Динамические массивы

using namespace std;

// Модель Ньютона: вычисляет температуру в момент времени
double modelTemperature(double T0, double Ts, double r, double t) {
    return Ts + (T0 - Ts) * exp(-r * t);
}

// Линейная аппроксимация методом наименьших квадратов
void linearFit(const vector<double>& x, const vector<double>& y, double& slope, double& intercept) {
    int n = x.size();
    double sumX = 0, sumY = 0, sumXY = 0, sumX2 = 0;

    for (int i = 0; i < n; ++i) {
        sumX  += x[i];
        sumY  += y[i];
        sumXY += x[i] * y[i];
        sumX2 += x[i] * x[i];
    }

    slope     = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX);
    intercept = (sumY - slope * sumX) / n;
}

// Расчёт коэффициента корреляции между двумя наборами данных
double correlationCoefficient(const vector<double>& x, const vector<double>& y) {
    int n = x.size();
    if (n == 0) return 0;

    double sumX = 0, sumY = 0;
    for (int i = 0; i < n; ++i) {
        sumX += x[i];
        sumY += y[i];
    }

    double avgX = sumX / n;
    double avgY = sumY / n;

    double numerator = 0, dx2 = 0, dy2 = 0;
    for (int i = 0; i < n; ++i) {
        double dx = x[i] - avgX;
        double dy = y[i] - avgY;
        numerator += dx * dy;
        dx2 += dx * dx;
        dy2 += dy * dy;
    }

    if (dx2 == 0 || dy2 == 0) return 0;
    return numerator / sqrt(dx2 * dy2);
}

// Безопасный ввод положительного числа
double readPositive(const string& label) {
    double value;
    while (true) {
        cout << label;
        cin >> value;
        if (cin.fail() || value <= 0) {
            cin.clear();
            cin.ignore(10000, '\n');
            cout << "Ошибка ввода. Повторите.\n";
        } else {
            return value;
        }
    }
}

int main() {
    // Переменные и массивы
    vector<double> timePoints, tempPoints;
    double T0, Ts, r, totalTime, dt;

    // Ввод исходных данных
    cout << "Начальная температура кофе (°C): ";
    cin >> T0;

    cout << "Температура окружающей среды (°C): ";
    cin >> Ts;

    r = readPositive("Коэффициент охлаждения (1/мин): ");
    totalTime = readPositive("Длительность моделирования (мин): ");
    dt = readPositive("Шаг времени (мин): ");

    // Заголовок таблицы
    cout << "\nМоделирование температуры кофе по закону Ньютона\n";
    cout << "----------------------------------------------\n";
    cout << setw(10) << "Время" << " | " << setw(15) << "Температура\n";
    cout << "----------------------------------------------\n";

    // Формат чисел
    cout << fixed << setprecision(2);

    // Расчёт температуры и сбор данных
    for (double t = 0; t <= totalTime; t += dt) {
        double T = modelTemperature(T0, Ts, r, t);
        timePoints.push_back(t);
        tempPoints.push_back(T);
        cout << setw(10) << t << " | " << setw(15) << T << "\n";
    }

    cout << "----------------------------------------------\n";

    // Аппроксимация прямой линии: T = a * t + b
    double a, b;
    linearFit(timePoints, tempPoints, a, b);

    // Расчёт корреляции
    double r_corr = correlationCoefficient(timePoints, tempPoints);

    // Вывод аппроксимирующей функции и корреляции
    cout << "\nЛинейная аппроксимация: T(t) = " << setprecision(4) << a << " * t + " << b << "\n";
    cout << "Коэффициент корреляции: r = " << r_corr << "\n";

    return 0;
}
